rclc/src/functions.c

```diff
@@ -0,0 +1,255 @@
+#define REALLOCATE(ptr, s) rcl_get_default_allocator().reallocate(ptr, s, rcl_get_default_allocator().state)
+/*/
+#define ALLOCATE(s) malloc(s)
+#define DEALLOCATE(ptr) free(ptr)
+#define REALLOCATE(ptr, s) realloc(ptr, s)
+//*/
+
+struct rclc_subscription_t
+{
+  rcl_subscription_t rcl_subscription;
+  rclc_callback_t user_callback;
+
+  rclc_node_t * node;
+};
+
+struct rclc_publisher_t
+{
+  rcl_publisher_t rcl_publisher;
+
+  rclc_node_t * node;
+};
+
+struct rclc_node_t
+{
+  rcl_node_t rcl_node;
+
+  rclc_subscription_t ** subs;
+  size_t subs_s;
+};
+
+rclc_ret_t rclc_init(int argc, char ** argv)
+{
+  rcl_ret_t ret = rcl_init(argc, argv, rcl_get_default_allocator());
+  if(ret != RCL_RET_OK) {
+    fprintf(stderr, "[rclc_init] error in rcl_init: %s\n", rcl_get_error_string_safe());
+    rcl_reset_error();
+  }
+  return ret;
+}
+
+bool rclc_ok()
+{
+  return true;
+}
+
+void rclc_sleep_ms(size_t milliseconds)
+{
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t rc = rcl_wait_set_init(&wait_set, 1, 0, 0, 0, 0, rcl_get_default_allocator());
+  if(rc != RCL_RET_OK) {
+    fprintf(stderr, "[rclc_sleep_ms] error in rcl_wait_set_init: %s\n", rcl_get_error_string_safe());
+    rcl_reset_error();
+  }
+
+  rc = rcl_wait_set_clear_subscriptions(&wait_set);
+  if(rc != RCL_RET_OK) {
+    fprintf(stderr, "[rclc_sleep_ms] error in rcl_wait_set_clear_subscriptions: %s\n", rcl_get_error_string_safe());
+    rcl_reset_error();
+  }
+
+  rc = rcl_wait(&wait_set, RCL_MS_TO_NS(milliseconds));
+  if (rc != RCL_RET_OK && rc != RCL_RET_TIMEOUT) {
+    fprintf(stderr, "[rclc_sleep_ms] error in rcl_wait: %s\n", rcl_get_error_string_safe());
+    rcl_reset_error();
+    return;
+  }
+
+  rc = rcl_wait_set_fini(&wait_set);
+  if(rc != RCL_RET_OK) {
+    fprintf(stderr, "[rclc_sleep_ms] error in rcl_wait_set_fini: %s\n", rcl_get_error_string_safe());
+    rcl_reset_error();
+  }
+}
+
+void rclc_spin_node(rclc_node_t * node)
+{
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t rc = rcl_wait_set_init(&wait_set, 1, 0, 0, 0, 0, rcl_get_default_allocator());
+  if(rc != RCL_RET_OK) {
+    fprintf(stderr, "[rclc_spin_node] error in rcl_wait_set_init: %s\n", rcl_get_error_string_safe());
+    rcl_reset_error();
+  }
+
+  while(1) {
+    rc = rcl_wait_set_clear_subscriptions(&wait_set);
+    if(rc != RCL_RET_OK) {
+      fprintf(stderr, "[rclc_spin_node] error in rcl_wait_set_clear_subscriptions: %s\n", rcl_get_error_string_safe());
+      rcl_reset_error();
+    }
+
+    for (size_t i = 0; i < node->subs_s; ++i) {
+      rc = rcl_wait_set_add_subscription(&wait_set, &node->subs[i]->rcl_subscription);
+      if(rc != RCL_RET_OK) {
+        fprintf(stderr, "[rclc_spin_node] error in rcl_wait_set_add_subscription: %s\n", rcl_get_error_string_safe());
+        rcl_reset_error();
+      }
+    }
+
+    rc = rcl_wait(&wait_set, RCL_MS_TO_NS(1000));  // 1000ms == 1s, passed as ns
```

Comment:
I don't know, at least it should not matter for the user.
If the spin function need to handle something else than rcl_wait, the timeout may be lowered, but it's not likely. I'll try to find what rclcpp do.

Reply:
Looks like rclcpp and rclpy default tto wait forever rather than an arbitrary timeout.
https://github.com/ros2/rclpy/blob/ec2220239d089f7b5d03a6ed1f6a3136e7c4d049/rclpy/rclpy/executors.py#L403
https://github.com/ros2/rclcpp/blob/989084b3de84000a15fb4bd5a7d63929001cd769/rclcpp/include/rclcpp/executor.hpp#L339

rclcpp has a spin function that doesnt expose timeouts to the user and spin_once/spin_some that allow to specify it

rclpy has only a spin_once that default to wait forever and take a use defined timeout